sinks/attribution/attribution.go

```diff
@@ -0,0 +1,346 @@
+package attribution
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/csv"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/axiomhq/hyperloglog"
+	"github.com/segmentio/fasthash/fnv1a"
+	"github.com/sirupsen/logrus"
+
+	veneur "github.com/stripe/veneur/v14"
+	"github.com/stripe/veneur/v14/samplers"
+	"github.com/stripe/veneur/v14/sinks"
+	"github.com/stripe/veneur/v14/ssf"
+	"github.com/stripe/veneur/v14/trace"
+	"github.com/stripe/veneur/v14/util"
+)
+
+// AttributionSinkConfig represents the YAML options that can be used to
+// configure an instance of AttributionSink
+type AttributionSinkConfig struct {
+	AWSAccessKeyID     util.StringSecret `yaml:"aws_access_key_id"`
+	AWSRegion          string            `yaml:"aws_region"`
+	AWSSecretAccessKey util.StringSecret `yaml:"aws_secret_access_key"`
+	HostnameTag        string            `yaml:"hostname_tag"`
+	SchemaVersion      int               `yaml:"schema_version"`
+	OwnerKey           string            `yaml:"owner_key"`
+	S3Bucket           string            `yaml:"s3_bucket"`
+	S3KeyPrefix        string            `yaml:"s3_key_prefix"`
+	VeneurInstanceID   string            `yaml:"veneur_instance_id"`
+}
+
+// ParseConfig decodes the map config for an S3 sink into an S3SinkConfig
+// struct.
+func ParseConfig(config interface{}) (veneur.MetricSinkConfig, error) {
+	attributionSinkConfig := AttributionSinkConfig{}
+	err := util.DecodeConfig(config, &attributionSinkConfig)
+	if err != nil {
+		return nil, err
+	}
+	return attributionSinkConfig, nil
+}
+
+// AttributionSink is a sink for flushing ownership/usage (attribution data) to S3
+type AttributionSink struct {
+	name             string
+	traceClient      *trace.Client
+	log              *logrus.Entry
+	s3KeyPrefix      string
+	veneurInstanceID string
+	s3Svc            s3iface.S3API
+	s3Bucket         string
+	attributionData  map[string]*TimeseriesGroup
+	ownerKey         string
+	commonDimensions map[string]string
+	hostnameTag      string
+	hostname         string
+	schemaVersion    int
+}
+
+// ErrS3ClientUninitializedError is an error returned when the S3 client
+// provided to AttributionSink is nil
+var ErrS3ClientUninitialized = errors.New("s3 client has not been initialized")
+
+// TimeseriesGroup represents an attributable unit of metrics data, complete with
+// owner, metadata corresponding to the metrics data grouping, and cardinality
+// tallying
+type TimeseriesGroup struct {
+	Name    string
+	Owner   string
+	Digests []uint32
+	Sketch  *hyperloglog.Sketch
+}
+
+// Create returns a pointer to an AttributionSink
+func Create(
+	server *veneur.Server, name string, logger *logrus.Entry,
```

Comment:
can we put each of these params on its own line so its easier to read

Reply:
done